prow/git/v2/client_factory.go

```diff
@@ -387,30 +395,46 @@ func (c *clientFactory) ClientForWithRepoOpts(org, repo string, repoOpts RepoOpt
 	}
 	gitMetrics.secondaryCloneDuration.WithLabelValues(org, repo).Observe((float64(time.Since(timeBeforeSecondaryClone).Seconds())))
 
-	// Here we do a targeted fetch, if the repoOpts for the secondary clone
-	// asked for it.
-	//
-	// TODO(listx): Change ensureCommits() so that it takes a "cacher" interface
-	// instead of a generic RepoClient. However, if ShareObjectsWithSourceRepo
-	// was not set, then we cannot run ensureCommits() with the cacher because
-	// the cacher will run on the primary, not secondary (and without --shared,
-	// this will have no effect on the secondary); we have to handle that case
-	// gracefully (perhaps with some logging also).
+	return repoClient, nil
+}
+
+func (c *clientFactory) ensureFreshPrimary(
+	cacheDir string,
+	cacheClientCacher cacher,
+	repoOpts RepoOpts,
+	org string,
+	repo string,
+) error {
+	if err := c.maybeCloneAndUpdatePrimary(cacheDir, cacheClientCacher, repoOpts); err != nil {
+		return err
+	}
+	// For targeted fetches by SHA objects, there's no need to hold a lock on
+	// the primary because it's safe to do so (git will first write to a
+	// temporary file and replace the file being written to, so if another git
+	// process already wrote to it, the worst case is that it will overwrite the
+	// file with the same data).  Targeted fetch. Only fetch those commits which
+	// we want, and only if they are missing.
 	if repoOpts.FetchCommits.Len() > 0 {
-		// Targeted fetch. Only fetch those commits which we want, and only
-		// if they are missing.
+		// Targeted fetch. Only fetch those commits which we want, and only if
+		// they are missing.
 		timeBeforeFetchBySha := time.Now()
-		if err := repoClientCloner.FetchCommits(repoOpts.NoFetchTags, repoOpts.FetchCommits.UnsortedList()); err != nil {
-			return nil, err
+		if err := cacheClientCacher.FetchCommits(repoOpts.NoFetchTags, repoOpts.FetchCommits.UnsortedList()); err != nil {
+			return err
 		}
 		gitMetrics.fetchByShaDuration.WithLabelValues(org, repo).Observe((float64(time.Since(timeBeforeFetchBySha).Seconds())))
 	}
```

Comment:
For this functionality to work as expected, I think this assumes that `repoOpts.ShareObjectsWithSourceRepo == true` which is not necessarily true at this point. If we're not sharing objects with the source repo then this needs to somehow update the secondary repo, not just the primary, right?

Reply:
No, because we are fetching new objects into the primary. It doesn't matter if we're sharing or not sharing across the primary or secondary because:

- if we're sharing, the secondary doesn't need to fetch the same objects (redundantly) again, because the primary has it already
- if we're not sharing, the secondary will get a full copy of everything in the (updated) primary that already has the objects

IOW the primary is *always created/refreshed first* before the secondary is created from it (either with a partial or non-partial (full) checkout).